captum/attr/_utils/lrp_rules.py

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env python3
+
+from abc import ABC, abstractmethod
+
+import torch
+
+
+class PropagationRule(ABC):
+    """
+    Base class for all propagation rule classes, also called Z-Rule.
+    STABILITY_FACTOR is used to assure that no zero divison occurs.
+    """
+
+    STABILITY_FACTOR = 1e-9
+
+    def forward_hook(self, module, inputs, outputs):
+        """Register backward hooks on input and output
+        tensors of linear layers in the model."""
+        self._handle_input_hooks = []
+        for input in inputs:
+            if not hasattr(input, "hook_registered"):
+                input_hook = self._create_backward_hook_input(input.data)
+                self._handle_input_hooks.append(input.register_hook(input_hook))
+                input.hook_registered = True
+        output_hook = self._create_backward_hook_output(outputs.data)
+        self._handle_output_hook = outputs.register_hook(output_hook)
+
+    @staticmethod
+    def backward_hook_activation(module, grad_input, grad_output):
+        """Backward hook to propagate relevance over non-linear activations."""
+        return grad_output
+
+    def _create_backward_hook_input(self, inputs):
+        def _backward_hook_input(grad):
+            relevance = grad * inputs
+            self.relevance_input = relevance.data
```

Comment:
Here we might be overwriting the relevance scores when the inputs is a tuple of multiple tensors ?

Reply:
Wow, that is correct. Is it enough to have `self.relevance_input` as list and append to it? It is only used for output of layerLRP.

Reply:
Usually we keep inputs and outputs as tuples if they consist of more than one tensor. To be precise we can have tuple only if we know that  `_backward_hook_input` will be called only for one input tensor ? 
 https://github.com/pytorch/captum/blob/95cf31104a64eb5f14aa2147852dd9bff885d72b/captum/attr/_utils/lrp_rules.py#L20